api/api.go

```diff
@@ -37,3 +45,165 @@ func ContainerOption(container container.Container) Option {
+	}
+
+	return services, nil
+}
+
+// StartService starts service serviceID.
+func (a *API) StartService(serviceID string) error {
+	sr, err := a.db.Get(serviceID)
+	if err != nil {
+		return err
+	}
+	sr, err = service.FromService(sr, service.ContainerOption(a.container))
+	if err != nil {
+		return err
+	}
+	_, err = sr.Start()
+	return err
+}
+
+// StopService stops service serviceID.
+func (a *API) StopService(serviceID string) error {
+	sr, err := a.db.Get(serviceID)
+	if err != nil {
+		return err
+	}
+	sr, err = service.FromService(sr, service.ContainerOption(a.container))
+	if err != nil {
+		return err
+	}
+	return sr.Stop()
+}
+
+// DeleteService stops and deletes service serviceID.
+// when deleteData is enabled, any persistent data that belongs to
+// the service and to its dependencies also will be deleted.
+func (a *API) DeleteService(serviceID string, deleteData bool) error {
+	s, err := a.db.Get(serviceID)
+	if err != nil {
+		return err
+	}
+	s, err = service.FromService(s, service.ContainerOption(a.container))
+	if err != nil {
+		return err
+	}
+	if err := s.Stop(); err != nil {
+		return err
+	}
+	// delete volumes first before the service. this way if
+	// deleting volumes fails, process can be retried by the user again
+	// because service still will be in the db.
+	if deleteData {
+		if err := s.DeleteVolumes(); err != nil {
+			return err
+		}
+	}
+	return a.db.Delete(serviceID)
+}
+
+// EmitEvent emits a MESG event eventKey with eventData for service token.
+func (a *API) EmitEvent(token, eventKey string, eventData map[string]interface{}) error {
+	s, err := a.db.Get(token)
+	if err != nil {
+		return err
+	}
+	event, err := event.Create(s, eventKey, eventData)
+	if err != nil {
+		return err
+	}
+	event.Publish()
+	return nil
+}
+
+// ExecuteTask executes a task tasKey with inputData and tags for service serviceID.
+func (a *API) ExecuteTask(serviceID, taskKey string, inputData map[string]interface{},
+	tags []string) (executionID string, err error) {
+	s, err := a.db.Get(serviceID)
+	if err != nil {
+		return "", err
+	}
+	s, err = service.FromService(s, service.ContainerOption(a.container))
+	if err != nil {
+		return "", err
+	}
+
+	// a task should be executed only if task's service is running.
+	status, err := s.Status()
+	if err != nil {
+		return "", err
+	}
+	if status != service.RUNNING {
+		return "", &NotRunningServiceError{ServiceID: s.Sid}
+	}
+
+	// execute the task.
+	eventID := uuid.NewV4().String()
+	exec, err := execution.New(s, eventID, taskKey, inputData, tags)
+	if err != nil {
+		return "", err
+	}
+	if err := exec.Execute(); err != nil {
+		return "", err
+	}
+	if err = a.execDB.Save(exec); err != nil {
+		return "", err
+	}
+	go pubsub.Publish(s.TaskSubscriptionChannel(), exec)
+	return exec.ID, nil
+}
+
+// ExecuteAndListen executes given task and listen for result.
+func (a *API) ExecuteAndListen(serviceID, task string, inputs map[string]interface{}) (*execution.Execution, error) {
```

Comment:
I'm not sure this is used anywhere and there is a big hack on the `time.Sleep` so I would recommend to remove it

Reply:
done